demos/graph-classification/supervised-graph-classification.ipynb

```diff
@@ -0,0 +1,559 @@
+{
```

Comment:
<p>The code in the nested loops is very dense; maybe there could be a little more whitespace  (e.g. before the inner <code>for</code>  and each of the <code>if</code>  statements.)</p><p> </p><p>Also, why clone model  when assigning to <code>best_model</code>? I don't think the <code>model</code> variable gets reused?</p><p> </p><p>Finally, if a user was trying to use this graph classification algorithm, I wonder if they'd struggle to translate this nested loop into a straight-line (non-loop) form that they'd actually use for creating and running the model. One option would be to a section demonstrating the conventional/basic code, and then go into the 10-fold validation code. What do you think?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1053/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Reply:
I'll update to make the loop easier to read (I hope.)

The copying of the model was added because I initially thought I can create one model and then re-initialise the weights at each iteration of the loop instead of creating a new one. Unfortunately, I could not find a way to reset a model's weights and was forced to create a new one every time. As you correctly point out, there is not need to copy the model anymore, and I will remove.

The training loop is rather common for k-fold training. Hopefully a small re-factoring will improve clarity. I don't want to assume that our users cannot un-roll a simple loop like this and I prefer not to make this notebook grow too long. That said, I had initially coded this using a single split of the data but because the data is small, every run would produce widely different performance and I thought not worth having it because it unfairly makes the method look poor.

Reply:
It's not a matter of "can" or "cannot", it's a matter of "how can we make it easy for them?". I know I personally appreciate things being spelled out when I'm learning them, even if I can work out the more complicated form. This is especially true because there's several layers here, not just loops: to understand the "core" concepts of doing graph classification, the user currently has to string together the definitions of `create_graph_classification_model`, `train_fold` and `get_generators` as well as detangle the 10-fold-related code, which is a lot more complicated than understanding the straight-line no-function form.

I mocked up a quick possible example of what an addition might look like https://nbviewer.jupyter.org/gist/huonw/2933706bc39adc2ba2b85b829b8513b2 (it can be downloaded in the top right hand corner).

I agree with your concern about the length, though, and it's unfortunate that every `fit` (and `evaluate`) call seems to be hitting #826, causing the amount of output to inflate. That said, along those lines, maybe we could have `verbose=0` for the evaluate call too to minimise the `1/1 [==============================] - 0s 8ms/step - loss: 0.4701 - acc: 0.7895` lines?